### ISSUE TITLE
Rich console produces garbled output on MinTTY when the console is less than 80 characters wide

### ISSUE THREAD

**Issue by adammurdoch** [gradle/gradle#8204]
The rich console hard-codes a console width of 80 characters when it cannot query the console metadata. This produces problems when the console is not exactly 80 characters wide.

- When the console is wider, the available space is not used and the user is shown less detail.
- When the console is narrower, the output is garbled.

When using MinTTY, Gradle is currently not able to query the console metadata. This should be possible, but it is not entirely clear how to just yet.

### Expected Behavior

The rich console uses the correct console width when using MinTTY.

### Current Behavior

The rich console uses the incorrect console width when using MinTTY.

**Comment by adammurdoch** [comment]
An alternative work around to the hard coded width would be to disable the dynamic content, but not the colour output, when we don't know the console width.

**Comment by dandosi** [comment]
Still happening for me in MinTTY/Cygwin64 on Win10 😢

Using the option/property --console=plain definitely.. "fixes" it, by having a plain output.
It seems tough for me to say when did this start happening, I think it may have been okay sometime ago with gradle 5.6.2,
but now it's bad for both gradle 5.6.2 or gradlew 6.1.1 (in the image)
![image](https://user-images.githubusercontent.com/1731479/74073849-a80a0980-49c0-11ea-87b9-763b6ea90182.png)

The console s/w version is:
mintty 3.1.0-1

**Comment by mkq** [comment]
For me too, with width > 80.
```
> uname -a
CYGWIN_NT-10.0 ... 3.1.4(0.340/5/3) 2020-02-19 08:49 x86_64 Cygwin

> cygcheck -c mintty
Cygwin Package Information
Package              Version        Status
mintty               3.1.4-1        OK

> stty size
63 237

> ./gradlew -Duser.home="$(cygpath -m ~)" --version

------------------------------------------------------------
Gradle 5.6.4
------------------------------------------------------------
...

> ./gradlew -Duser.home="$(cygpath -m ~)" tasks
Starting a Gradle Daemon, 1 incompatible Daemon could not be reused, use --status for details
                                                               <-------------> 0% INITIALIZING [30ms]                          <-------------> 0% INITIALIZING [112ms]
> Starting Daemon                        > IDLE           <-------------> 0% INITIALIZING [313ms]                          <-------------> 0% INITIALIZING [413ms]                          <-------------> 0% INITIALIZING [53
                <-------------> 0% INITIALIZING [611ms]                          <-------------> 0% INITIALIZING [743ms]                          <-------------> 0% INITIALIZING [811ms]                          <-----------
m                                      <-------------> 0% INITIALIZING [1s]
            <-------------> 0% INITIALIZING [2s]                          <-------------> 0% INITIALIZING [3s]lding buildSrc                        > Building buildSrc > Loading projects                    > Building buildSrc
                                                                                                                                        > Building buildSrc > :buildSrc                    > Building buildSrc
                                                       > Building buildSrc > :buildSrc:compileJava                    > Building buildSrc > :buildSrc:compileJava > Resolve dependencies of :buildSrc:compileClasspath
m                                                          > Building buildSrc > :buildSrc:compileJava > Resolve files of :buildSrc:compi<-------------> 0% INITIALIZING [6s] Building buildSrc > :buildSrc:compileJava
A ><-------------> 0% CONFIGURING [6s]
1B > root project                         <-------------> 0% CONFIGURING [7s]                         ><-------------> 0% CONFIGURING [8s]s of :classpath           <-------------> 0% CONFIGURING [9s]iles of :classpath        <
[1B > root project                                                               <-------------> 0% EXECUTING [9s]
> Task :tasks

------------------------------------------------------------
Tasks runnable from root project
...
```

I usually have `-Duser.home=... -Djava.io.tmpdir=C:/tmp/java` in environment variable `JAVA_TOOL_OPTIONS`, but unset it for this test, as it prints message `Picked up JAVA_TOOL_OPTIONS: ...`. Comment https://github.com/gradle/gradle/issues/6874#issuecomment-432111090 suggested this might be a problem.

Also, gradle seems to insert hard line breaks, for example in JUnit test outputs when shown with `--info`. This is really annoying when copying or doing something with that output via mintty's user commands with variable `MINTTY_OUTPUT`. Should I open another issue for that?

**Comment by mkq** [comment]
According to PR 12844, I tested with a different `TERM` environment variable. With a white background, I get some unreadable white on white messages, e.g. this part:
```
> JAVA_TOOL_OPTIONS= TERM=mintty ./gradlew -Duser.home="$(cygpath -m ~)" tasks

> Task :tasks

------------------------------------------------------------
Tasks runnable from root project
------------------------------------------------------------

Bpplication tasks
-----------------
assemble - Assembles the outputs of this project.
```
Also, somehow "Application" became "Bpplication" when copying.

There is also some duplicated progress info: This is a single invocation (at 150 char width):
```
> JAVA_TOOL_OPTIONS= TERM=mintty ./gradlew -Duser.home="$(cygpath -m ~)" --version
Downloading https://services.gradle.org/distributions/gradle -6.3-bin.zip
.........10%..........20%..........30%.........40%..........50%..........60%.........70%..........80%..........90%..........100%
                                                                                                                                                      > JAVA_TOOL_OPTIONS= TERM=mintty ./gradlew -Duser.home="$(cygpath -m ~)" --version
Downloading https://services.gradle.org/distributions/gradle -6.3-bin.zip
.........10%..........20%..........30%.........40%..........50%..........60%.........70%..........80%..........90%..........100%

Welcome to Gradle 6.3!
…
```
(Spaces in `gradle -6.3` added by me to avoid linking with another issue.)

This was with a simplified prompt. With my usual two-line prompt including mintty's [scroll markers](https://github.com/mintty/mintty/wiki/CtrlSeqs#scroll-markers) and zsh's `setopt prompt_subst` and much more (both with black and white background), there's an additional problem: My prompt line gets moved:
![image](https://user-images.githubusercontent.com/1584258/80475425-25eb0a80-8949-11ea-89ed-2d927214dc87.png)
For reference, this is the correct prompt after invoking another command (and the first line of the next prompt):
![image](https://user-images.githubusercontent.com/1584258/80475242-e15f6f00-8948-11ea-90df-5d44a0572de5.png)

**Comment by mkq** [comment]
Another observation: `--console=plain` does not get rid of control characters. I tried `TERM=dumb ./gradlew.bat --console=plain --version` as a compilation command in GNU Emacs, and it looks like this:
![image](https://user-images.githubusercontent.com/1584258/81719662-9e8bb400-947d-11ea-99df-ad8a2f6a9710.png)

**Comment by mkq** [comment]
TL;DR: Found a **workaround**: Set **environment variable `CYGWIN=disable_pcon`** globally. (Or, for those that have already set that variable, append ` disable_pcon` to its value.)

Long version:
> Also, gradle seems to insert hard line breaks, for example in JUnit test outputs when shown with `--info`. This is really annoying when copying or doing something with that output via mintty's user commands with variable `MINTTY_OUTPUT`. Should I open another issue for that?

I noticed another Java program having the same issue of hard line wraps, so I looked for mintty issues and found [mintty issue 971](https://github.com/mintty/mintty/issues/971). That issue links to the [cygwin mailing list](https://cygwin.com/ml/cygwin/2020-03/msg00088.html), where [a reply](https://sourceware.org/legacy-ml/cygwin/2020-03/msg00094.html) mentions setting environment variable `CYGWIN=disable_pcon`.
That fixed it for me, and not only the line breaks, but also the "garbling". I need to set it early on though, for example setting it on a command-by-command basis with `env` did not help, so I put it in the Windows environment (control panel).

**Comment by guipivoto** [comment]
@mkq Thanks for sharing this workaround.
I tested and it worked for me too.. As you said, it works only if you set it as a windows environment variable (I think you could add this to the TL;DR).

My environment:
Minty: 3.1.7
Gradle: 5.4.1

**Comment by tglaeser** [comment]
At least in my environment this issue happens for any character wide console nor would think that an 80 character wide console is usable anyway; also the proposed workaround makes no difference:
```
$ mintty --version
mintty 3.2.0 (x86_64-pc-msys)
© 2013/2020 Andy Koppe / Thomas Wolff
License GPLv3+: GNU GPL version 3 or later
There is no warranty, to the extent permitted by law.

$ gradle --version
------------------------------------------------------------
Gradle 6.6-rc-3
------------------------------------------------------------

Build time:   2020-07-24 14:04:09 UTC
Revision:     13b5dcc4f294b395e61d937a3601fd47be38e76e

Kotlin:       1.3.72
Groovy:       2.5.12
Ant:          Apache Ant(TM) version 1.10.8 compiled on May 10 2020
JVM:          1.8.0_262 (Azul Systems, Inc. 25.262-b19)
OS:           Windows 10 10.0 amd64
```
With everything else unchanged, I have another project using Gradle version `5.0` where
```
./gradlew --console=rich
```
works just fine. What are the plans for the `6.6` release to have this addressed?

**Comment by mkq** [comment]
After a gradle update from 6.4 to 6.7 (and with `CYGWIN=… disable_pcon` still set globally in Windows control panel as described [above](https://github.com/gradle/gradle/issues/8204#issuecomment-644232747)), the output is broken again:

```
> ./gradlew.BAT assemble
Picked up JAVA_TOOL_OPTIONS: -Duser.home=C:/home/XXX -Djava.io.tmpdir=C:/tmp/java
Starting a Gradle Daemon (subsequent builds will be faster)


> Starting Daemon> IDLE<-------------> 0% INITIALIZING [89ms]> Running init scripts<-------------> 0% INITIALIZING [190ms]<-------------> 0% INITIALIZING [289ms]<-------------> 0% INITIALIZING [389ms]<-------------> 0% INITIALIZING [489ms]<-------------> 0% INITIALIZING [590ms]<-------------> 0% INITIALIZING [690ms]> Running init scripts > Compiling initialization script …
```

**Comment by doctorpangloss** [comment]
Also experiencing this issue in 6.7. It appears also in `./gradlew` i.e. the bash script, since I use bash as provided by Git's installer.

Prefixing commands with `TERM=mintty` works around the issue.

**Comment by mkq** [comment]
Thanks, but it does not work for me: With gradle 6.7 (and `CYGWIN=winsymlinks:native detect_bloda disable_pcon`): 

```
> TERM=mintty ./gradlew --stop; TERM=mintty ./gradlew clean
Picked up JAVA_TOOL_OPTIONS: -Duser.home=C:/home/xxx -Djava.io.tmpdir=C:/tmp/java


> Connecting to DaemonStopping Daemon(s)

> IDLE1 Daemon stopped

> IDLEPicked up JAVA_TOOL_OPTIONS: -Duser.home=C:/home/xxx -Djava.io.tmpdir=C:/tmp/java
Starting a Gradle Daemon, 3 stopped Daemons could not be reused, use --status for details


> Starting Daemon> IDLE<-------------> 0% INITIALIZING [39ms]> Running init scripts<-------------> 0% INITIALIZING [139ms]<-------------> 0% INITIALIZING [239ms]<-------------> 0% INITIALIZING [340ms]<-------------> 0% INITIALIZING [439ms]<-------------> 0% INITIALIZING [539ms]<-------------> 0% INITIALIZING [640ms]<-------------> 0% INITIALIZING [740ms]<-------------> 0% INITIALIZING [839ms]> Evaluating settings<-------------> 0% CONFIGURING [940ms]> Loading projects<-------------> 0% CONFIGURING [1s]> IDLE> root project<=============> 100% CONFIGURING [1s]> IDLE<-------------> 0% EXECUTING [1s]> :clean
Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.7/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 6s
1 actionable task: 1 up-to-date


<-------------> 0% WAITING> :clean
> 
```

**Comment by doctorpangloss** [comment]
To clarify I use mintty through git bash. You should use a console / terminal and search online what to use for TERM.

**Comment by mkq** [comment]
I noticed that running `gradlew.bat` inside [winpty](https://github.com/rprichard/winpty) looks good, both with and without `disable_pcon`, except it uses inverse colors. I guess that's worse than `--console=plain` for my taste, though.

**Comment by mkq** [comment]
still broken in Gradle 7.3.3

**Comment by doctorpangloss** [comment]
For users of `git-bash` on Windows Terminal, add `alias gradle="TERM=mintty gradle"` to your `.bash_profile`

**Comment by sschuberth** [comment]
Probably this is blocked by https://github.com/gradle/native-platform/issues/33.
